plenum/test/consensus/test_three_pc_validator.py

```diff
@@ -0,0 +1,174 @@
+        assert validator.validate(msg) == (STASH_CATCH_UP, CATCHING_UP)
+
+
+def test_outside_watermark(validator, view_no, inst_id, pp_seq_no):
+    H = 10
+    validator._data.is_participating = True
+    validator._data.low_watermark = 0
+    validator._data.high_watermark = H
+    for msg in create_3pc_msgs(view_no=view_no,
+                               pp_seq_no=pp_seq_no + H,
+                               inst_id=inst_id):
+        assert validator.validate(msg) == (STASH_WATERMARKS, OUTSIDE_WATERMARKS)
+
+
+def test_not_commit_with_old_view(validator, view_no, inst_id, pp_seq_no):
+    validator._data.view_no = view_no + 1
+    validator._data.pp_seq_no = pp_seq_no
+    pp, p, _ = create_3pc_msgs(view_no=view_no,
+                               pp_seq_no=pp_seq_no,
+                               inst_id=inst_id)
+    assert validator.validate(pp) == (DISCARD, OLD_VIEW)
+    assert validator.validate(p) == (DISCARD, OLD_VIEW)
+
+
+def test_commit_old_view_vc_in_progress(validator, view_no, inst_id, pp_seq_no):
+    validator._data.view_no = view_no + 1
+    validator._data.pp_seq_no = pp_seq_no
+    validator._data.legacy_vc_in_progress = False
+    _, _, commit = create_3pc_msgs(view_no=view_no,
+                                   pp_seq_no=pp_seq_no,
+                                   inst_id=inst_id)
+    assert validator.validate(commit) == (DISCARD, OLD_VIEW)
+
+
+def test_commit_legacy_last_prepared_sertificate_is_none(validator, view_no, inst_id, pp_seq_no):
+    validator._data.view_no = view_no + 1
+    validator._data.pp_seq_no = pp_seq_no
+    validator._data.legacy_vc_in_progress = True
+    validator._data.legacy_last_prepared_before_view_change = None
+    _, _, commit = create_3pc_msgs(view_no=view_no,
+                                   pp_seq_no=pp_seq_no,
+                                   inst_id=inst_id)
+    assert validator.validate(commit) == (DISCARD, OLD_VIEW)
+
+
+def test_commit_greater_then_legacy_last_prepared_sertificate(validator, view_no, inst_id, pp_seq_no):
+    validator._data.view_no = view_no + 1
+    validator._data.pp_seq_no = pp_seq_no
+    validator._data.legacy_vc_in_progress = True
+    validator._data.legacy_last_prepared_before_view_change = (view_no, pp_seq_no - 1)
+    _, _, commit = create_3pc_msgs(view_no=view_no,
+                                   pp_seq_no=pp_seq_no,
+                                   inst_id=inst_id)
+    assert validator.validate(commit) == (DISCARD, GREATER_PREP_CERT)
+
+
+def test_process_if_synced_and_vc_in_progress(validator, view_no, inst_id, pp_seq_no):
+    validator._data.view_no = view_no + 1
+    validator._data.node_mode = Mode.synced
+    validator._data.legacy_vc_in_progress = True
+    validator._data.legacy_last_prepared_before_view_change = (view_no, pp_seq_no + 1)
+    _, _, commit = create_3pc_msgs(view_no=view_no,
+                                   pp_seq_no=pp_seq_no + 1,
+                                   inst_id=inst_id)
+
```

Comment:
The following tests needs to be migrated to OrderingService as well:
- plenum/test/replica/test_api.py
- plenum/test/replica/test_buffers_cleaning.py
- plenum/test/replica/test_create_3pc_batch.py
- plenum/test/replica/test_update_watermarks_api.py
- plenum/test/freshness/test_replica_freshness.py
- plenum/test/node_request/test_pre_prepare/test_pp_obsolescence.py